### PR TITLE
File picker and bookmarks API

### LINKED ISSUE #7209: OpenFolderDialog - Support multi selection
**Is your feature request related to a problem? Please describe.**
The native OpenFolderDialog control currently only supports single selection. At least on Windows it could support multi selection.

**Describe the solution you'd like**
Just like OpenFileDialog support multi selection

**Describe alternatives you've considered**

**Additional context**

### LINKED ISSUE #7234: New file picker API
Our current file dialogs are desktop-centric and assume that the app has direct access to the file system, which is not the case for sandboxed environments.

While keeping the current API for desktop platforms, we need to design a new one that would work well with WebAssembly, macOS App Store apps, iOS and Android (possibly with UWP too, but I'm not sure how relevant it is since we don't support XBox and HoloLens anyway).

Main differences of the API would be:
1) the lack of direct access to the file system, so the API should provide `Stream`s instead of file paths
2) apps might need to implement extra functionality to conform with the OS guidelines (e. g. subscribe to file changes via [NSFilePresenter](https://developer.apple.com/documentation/foundation/nsfilepresenter?language=objc) in case of iOS
2) directory access would be more complicated (e. g. https://developer.apple.com/documentation/uikit/view_controllers/providing_access_to_directories?language=objc), so we might want to skip it in the initial implementation
3) some sandboxes allow to save a file "bookmark" that allows the app to access the same file after restart (see https://github.com/AvaloniaUI/Avalonia/pull/6540)
4) more complicated required file type information rather than simply a file extension. For example iOS requires to specify a [uniform type identifier](https://developer.apple.com/documentation/uniformtypeidentifiers?language=objc) that are not MIME-types (e. g. `com.adobe.pdf`, `org.idpf.epub-container`, `public.mp3`
5) sometimes it's only possible to "export" a file. In case of WebAssembly saving file means creating an object URI from ReadableStream-backed Response object and waiting for browser to start a download), we can only report the intended file name and file type via a virtual `Content-Disposition` header. In case of iOS that would be the Share function
6) Some platforms (e. g. iOS) has different APIs for accessing documents (UIDocumentPickerViewController) and photos/videos (UIImagePickerController)

```cs
class FilePickerOpenOptions
{
    string Title { get; set; }
    string DefaultFileName { get; set; }
    IList<FilePickerFileType> FileTypes { get; set; }
}

class FilePickerOpenOptions
{
    string Title { get; set; }
    string DefaultFileName { get; set; }
    FilePickerFileType DefaultFileType {get; set; }
    IList<FilePickerFileType> FileTypes { get; set; }
}


class FilePickerFileType
{
    // For desktop
    string Title { get; init; }
    IReadOnlyList<string> Extensions {get; init; }
    // For web
    IReadOnlyList<string> MimeTypes {get; init; }
    // For Apple platforms
    IReadOnlyList<string> AppleUniformTypeIdentifiers {get; init; }

}

// We should provide some built-in types for common file types
class FilePickerFileTypes
{
     static FilePickerFileType Pdf {get;} = new FilePickerFileType
     {
        Title = "PDF document",
        Extensions = new [] {"pdf"},
        AppleUniformTypeIdentifiers = new[] {"com.adobe.pdf"},
        MimeTypes = new [] { "application/pdf" }
     };
}

interface IFilePicker
{
    bool CanOpen { get; }
    Task<IFilePickerFile> OpenAsync(FilePickerOptions options);
    bool CanSave { get; }
    Task<IFilePickerFile> SaveAsync(FilePickerOptions options);
    bool CanExport { get; }
    Task Export(FilePickerOptions options, Func<Stream, FilePickerFileType, Task> writer);
    IFilePickerBookmark OpenBookmark(string bookmark);
}

interface IFilePickerWriteContext
{
    Stream Stream {get;}
    FilePickerFileType FileType {get;}
}

interface IFilePickerBookmark : IDisposable
{
    bool CanOpenRead {get;}
    bool CanOpenWrite {get;}
    Task<Stream> OpenRead();
    Task<Stream> OpenWrite();
}

interface IFilePickerFile : IDisposable
{
     Stream Stream {get;}
     bool CanBookmark{get;}
     Task<string> SaveBookmark();
}

```

### PR DESCRIPTION
**Pull request by maxkatz6**
## What does the pull request do?

Initial implementation of new file picker API.
API layer was also inspired by HTML5 browser File API and naming from UWP. 
Xamarin Essentials, Qt, Electron and react-native-document-picker were used as a reference for implementation (aside from native API docs).

Current progress:
- [x] Open/save file and open folder dialogs API
- [x] Bookmark API (without required additions for WASM)
- [x] Implementation for main desktop OSes - win, osx, lin
- [x] iOS and Android implementation
- [x] Update managed file dialog to work with new interfaces
- [x] Folder picker support for Android.
- [x] FileAPI Browser support
- [ ] Finalize the API and set it in stone

Out of this PR, but to be done in this release:
- [ ] Fallback browser support with [a ponyfil](https://github.com/jimmywarting/native-file-system-adapter)
- [ ] macOS sandbox safe bookmarks implementation (https://github.com/AvaloniaUI/Avalonia/pull/6540)
- [ ] Export API. Or Share API, naming? I would prefer to keep it minimal. Required for iOS and WASM support, as there is no "Save" pickers (except newest new File API).

Potentially in the future:
- [ ] StorageFile file type

Usage:
```csharp
var folder = await window.StorageProvider.OpenFolderPickerAsync(new FolderPickerOpenOptions());
var openedFile = await window.StorageProvider.OpenFilePickerAsync(new FilePickerOpenOptions()
{
    SuggestedStartLocation = folder
});

// Open file to write.
var stream = await openedFile.OpenWrite();

// To save in local database.
var bookmarkId = openedFile.SaveBookmark();

var file = await window.StorageProvider.OpenFileBookmarkAsync(bookmarkId);

// Uri as it can be "file:" or "content:" or relative link depending on the platform. Following Qt API.
if (file.TryGetUri(out var uri) && uri.IsAbsoluteUri)
{
    var absolutePath = uri.LocalPath;
}
```

## Checklist

- [ ] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Breaking changes
None

## Obsoletions / Deprecations
SystemDialog and OpenFileDialog/SaveFileDialog/OpenFolderDialog are obsolete now.

## Fixed issues
Fixes https://github.com/AvaloniaUI/Avalonia/issues/7234
Fixes https://github.com/AvaloniaUI/Avalonia/issues/7209

## Questions 

Q: Should I rename `topLevel.StorageProvider` to `topLevel.FilePicker` with related interface names? Or `FileSystem`? Or `FileManager` to match other `*Manager` classes.
Q: Should API be change in order to allow opening dialogs without window/toplevel at all? It's possible only on desktop.
Q: Replace bool properties with enum flags?

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0021086-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by robloo** [comment]
Nice!

**Comment by Symbai** [comment]
Does it finally support multi folder selection? #7209

**Comment by maxkatz6** [comment]
No

**Comment by Symbai** [comment]
If you rewrite the whole API and obsolete existing API, wouldn't it make sense to look at existing issues and make sure they don't exist in the new one too? When, if not now, is the best chance for that?

**Comment by maxkatz6** [comment]
I don't think I have seen anywhere multi-folder picker API in xplat frameworks. And overall usefulness seems low.
But I will look which platforms actually support it. If it's more than one or two, than it's possible to do now.

**Comment by Symbai** [comment]
No idea what xplat frameworks is. But multi folder picker is a common thing on Windows and MacOS and Linux (basically all desktop environments). Winforms support it therefore also WPF. [WinUI currently seem to not support it](https://github.com/microsoft/microsoft-ui-xaml/issues/2854#issuecomment-655260876), because it comes from UWP which also hasn't supported it but it [was also requested](https://stackoverflow.com/questions/52413048/selecting-multiple-folders-async-in-uwp-with-c-sharp). For mobile it might not make much sense but I have never used a mobile app that asked me to select a folder at all, because working with folders on mobile devices is quite uncommon.

**Comment by timunie** [comment]
I really think we will need some docs how to handle this once it is ready. What do you think @maxkatz6 ?

**Comment by maxkatz6** [comment]
File picker with bookmarks (reusing same file without opening picker again) demo:

https://user-images.githubusercontent.com/3163374/174466294-456b73fe-0870-4dc1-987f-fc36b3cfcca4.mp4

**Comment by maxkatz6** [comment]
This PR can now be tested, if anybody wants to try.

I have to update macOS implementation to support file types. And test macos/ios once more.
And probably I can try to make "default name with extension" behavior more consistent across platforms in this PR. Somewhere extension is appended automatically, somewhere it's not.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0021227-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0021285-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

File: samples/ControlCatalog/Pages/NumericUpDownPage.xaml [hunk 1]
```diff
@@ -76,21 +76,21 @@
       <StackPanel Orientation="Vertical" Margin="10">
         <Label Target="upDown" FontSize="14" FontWeight="Bold" VerticalAlignment="Center">Usage of decimal NumericUpDown:</Label>
         <NumericUpDown Name="upDown" Minimum="0" Maximum="10" Increment="0.5"
-                       CultureInfo="en-US" VerticalAlignment="Center" Value="{Binding DecimalValue}"
```
Comment (maxkatz6) [inline samples/ControlCatalog/Pages/NumericUpDownPage.xaml:79]: Removed default culture info, as it was crashing control catalog in the browser.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0021348-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0021548-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0021619-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Review by kekekeks — APPROVED** [review]

File: src/Avalonia.Base/Platform/Storage/FilePickerSaveOptions.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+{
+    /// <summary>
+    /// Gets or sets the text that appears in the title bar of a file dialog.
+    /// </summary>
+    public string? Title { get; set; }
+
+    /// <summary>
+    /// Gets or sets the file name that the file save picker suggests to the user.
+    /// </summary>
+    public string? SuggestedFileName { get; set; }
```
[the remark this answers is not in the document]
Comment (maxkatz6) [reply src/Avalonia.Base/Platform/Storage/FilePickerSaveOptions.cs:18]: For "open" dialog it should be "SelectedFile" or something of IStorageFile type. We can add it later on top of that api.

File: src/Avalonia.Base/Platform/Storage/FilePickerSaveOptions.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    /// <summary>
+    /// Gets or sets the default extension to be used to save the file.
+    /// </summary>
+    public string? DefaultExtension { get; set; }
+
+    /// <summary>
+    /// Gets or sets the initial location where the file open picker looks for files to present to the user.
+    /// </summary>
+    public IStorageFolder? SuggestedStartLocation { get; set; }
```
[the remark this answers is not in the document]
Comment (maxkatz6) [reply src/Avalonia.Base/Platform/Storage/FilePickerSaveOptions.cs:28]: Yep

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0021621-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Pull request merged**